teps/0138-decouple-api-and-feature-versioning.md

```diff
@@ -72,8 +91,329 @@ Per [TEP-0033](https://github.com/tektoncd/community/blob/main/teps/0033-tekton-
+
+However, the default value of `enable-api-fields` will **not** be changed and it will remain to be `beta`, so the **beta** features will continue to be enabled by default.
+
+- **Impacts on users:**
+    - Cluster operators:
+      - Current cluster operators with `enable-api-fields` set to `alpha` or `beta` should not experience any changes.
+      - This makes it possible for cluster operators to have full control over only stable feature usages, rather than user overrides. For example, currently if cluster operators want their users to only opt-in “stable” features, they cannot do so for v1beta1 apiVersion for the exception of resolver, object params and results.
+    - Pipeline and Task authors: 
+    This affects pipeline users in the situation where cluster operators have chosen to set `enable-api-fields` to `stable` (i.e. those who have changed the current default `beta` value) and who are accidentally using beta features, like resolvers, in v1beta1.
+
+### Per feature flag for _new_ api-driven features
+Introduce per-feature flags for each **new** API driven feature. Each feature will have its own flag, instead of using the group API driven flag `enable-api-fields` to enable or disable all features of a stability level. Note that our proposal only takes effect on api-driven features while behavioural flags will remain the existing beahviour.
+
+The flag will also include the maintainer information and stability level for the feature as the source of truth. New behavioural features will also have a new per-feature flag to either enable or disable the feature. This will allow behavioural features that has values leading to behaviors in different stability levels to be turned on or off instead of depending on the stability level of features that are enabled.
+
+For example:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: feature-flags
+data:
+  # alpha: v0.53
+  #
+  # Pipeline in pipeline has ... functionalities.
+  # It is disabled by default.
+  enable-pipeline-in-pipeline: "false"
+  # alpha: v0.53
+  # 
+  # Trusted artifacts has ... functionalities.
+  # It is disabled by default.
+  enable-trusted-artifacts: "true"
+```
+
+See [implementation plan](#implementation-plan) for more details on the PerFeatureFlag struct.
+
+All **new** features can only be enabled via per-feature flags. When they first get introduced as alpha, they will be disabled by default. When new features get promoted to stable, they will be enabled by default according to the following table:
+
+| Feature stability level | Default  |
+| ----------------------- | -------- |
+| Stable                  | Enabled; Cannot be disabled  |
+| Beta                    | Disabled |
+| Alpha                   | Disabled |
+
+Note that for per feature flag that has stabilized, they cannot be disabled and we will remove the feature flag after 3 releases after it has become stable. Notification will be given to users via release notes.
```

Comment:
Not sure if it's covered by our API rules of 9 months, but by releases here, do we mean LTS or any releases ? and if any, how do we take LTS into account ? 

In addition, the flag "might" be removed (aka being there or not shouldn't have an effect, nor its value) but I wonder if we should "validate" thoses ? (aka fail if a feature-flag is no more)

Reply:
I agree that the removal of the flag should have no effect on the 'stable' feature usage since they cannot be disabled in any way. 

> but I wonder if we should "validate" thoses
I would like to clarify, by "validate" you mean validate features with their per feature flag at "stable" level or do you mean we should validate them after the flags are removed? If it is the first one, implementation-wise, we have added a fields, "LockToDefault" (similar to what k8s does now), indicating that we might not even need to "validate" stable features since they are locked to be turned on by default.

> Not sure if it's covered by our API rules of 9 months, but by releases here, do we mean LTS or any releases ?
That's a great point. I intended to mean 3 'general' releases (either LTS or non-LTS). The idea here is any 3 releases shall include 1 LTS, so at least one feature remains stable in 1 LTS the feature flag could be removed. 

In terms of API compatibility policies, I think this is exactly the new part and we would need to reach agreement on. I would lean towards that this doesn't need to fall under the 9 months' deprecation rule since it does not affect users in any case. Happy to hear from your opinions on this.

Reply:
Let's say a feature is marked as stable in `0.53`, the flag can potentially be deleted in `0.57` after three releases (`0.54` LTS, `0.55`, and `0.56`). Now, the next LTS is `0.58`. If a cluster is upgraded from `0.54` to `0.58`, do we expect the cluster operator to delete that feature flag from their deployment script?

Reply:
If an operator enables a feature `foobar` before it's stable, they will have the `enable-foobar` flag set to true. Once the feature is promoted to stable, they should be able to continue using the feature seamlessly, i.e. we should ignore `enable-foobar=true`.
 
If an operator disables a feature `foobar` before it's stable, once the feature becomes stable they will have to decide what to do. If they still don't want users to use that feature they may employ a custom admission controller to do so. I believe the best user experience would be to be notified that the flag is deprecated and will be removed in X releases so that they have time to develop an alternate solution - that requires the flag to continue work for some time after the flag reaches `stable`. 

If an operator does not set the `enable-foobar` flag (they use the default value), the switch to `stable` will be transparent to them. They will be aware through release notes that a new API feature is available.

I don't think we should keep the flag around if it cannot be used. I would recommend deprecating it instead and then removing it after the deprecation period.

> do we expect the cluster operator to delete that feature flag from their deployment script

@pritidesai I think we should issue a warning if the flag is set to `false` (i.e. feature disabled) starting with `0.53`. From `0.58` we can apply the behaviour we apply for unknown flags (I'm not sure if we fail or ignore?)

Reply:
I agree with the suggestion that we are deprecating stabilized feature flags and removing it eventually. This shall fully comply with our API policies and giving users enough notice from release notes to react.